### DISCUSSION TITLE
4.19.1.1: `rpmbuild --undefine foo` is not working

### DISCUSSION THREAD

**Discussion by kloczek** [rpm-software-management/rpm#2939] [Q&A] [answered]
I know that this option is not mentioned anywhere in documentation however IMO it should be present to be able not only define from command line macro using `--define foo` but as well undefine it if it is present in global or spec macros.
So this is I've decided to raise this not as RFE but issue ticket.

More interesting is that on use `rpmbuild --undefine foo` rpmbuild do not fails with message informing that it is is unsupported option.
I'm not 100% how `--undefine` option is recognised but looks like some parts or rpm command line parsing are involved.
I made two test specs to expose that
<details>
<summary>test-define.spec</summary>

```spec
Summary:        test --define
Name:           test-define
Version:        1
License:        foo
Release:        2%{?dist}

%description
Test --define.

%prep

%if 0%{?test}
echo "test defined"
%endif

%check
echo "%{name} package has no test suite"

%files
```
</details>

<details>
<summary>test-undefine.spec</summary>

```spec
%define         test    1

Summary:        test --undefine
Name:           test-undefine
Version:        1
License:        foo
Release:        2%{?dist}

%description
Test --undefine.

%prep

%if 0%{?test}
echo "test defined"
%endif

%check
echo "%{name} package has no test suite"

%files
```
</details>

And here are results:
```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bp test-define.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.ySytzr
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0

RPM build warnings:
    source_date_epoch_from_changelog set but %changelog is missing
[tkloczko@pers-jacek SPECS]$ rpmbuild -bp test-define.spec --define "test 1"
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.PANGwI
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ echo 'test defined'
test defined
+ RPM_EC=0
++ jobs -p
+ exit 0

RPM build warnings:
    source_date_epoch_from_changelog set but %changelog is missing
```
So it works as expected.
And here is `--undefine` test
```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bp test-undefine.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.GM3NF2
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ echo 'test defined'
test defined
+ RPM_EC=0
++ jobs -p
+ exit 0

RPM build warnings:
    source_date_epoch_from_changelog set but %changelog is missing
[tkloczko@pers-jacek SPECS]$ rpmbuild -bp test-undefine.spec --undefine test
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.2kdiZT
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ echo 'test defined'
test defined
+ RPM_EC=0
++ jobs -p
+ exit 0

RPM build warnings:
    source_date_epoch_from_changelog set but %changelog is missing
```
So `--undefine` as cmd param is recognised however second case is not handled as expected.
I'm almost sure that fixing that could be one one or two code lines change .. however cannot locate that.
May I ask for help? 🤔

**Comment by pmatilai** [comment, marked as answer]
--define and --undefine execute long before the spec is parsed, so while they can affect the way a spec is parsed, neither can affect a macro defined inside a spec.

**Comment by kloczek** [comment]
> --define and --undefine execute long before the spec is parsed, so while they can affect the way a spec is parsed, neither can affect a macro defined inside a spec.

Please have look one more time on my examples.
`--define` IS working and at the same time `--undefine` IS NOT working so it cannot be matter of when something is parsed.

**Comment by pmatilai** [comment]
No, read what I wrote more carefully. Neither --define or --undefine can affect a macro *defined inside a spec*.
Both can affect the parse, and cause things to be defined or undefined, but only wrt macros that are defined before the macro is parsed.

**Comment by kloczek** [comment]
> No, read what I wrote more carefully. Neither --define or --undefine can affect a macro _defined inside a spec_.
> Both can affect the parse, and cause things to be defined or undefined, but only wrt macros that are defined before the macro is parsed.

Please have look ONE MORE time on specs example which disproves what you just wrote.

**Reply by pmatilai** [reply]
No, it doesn't.

**Comment by kloczek** [comment]
> No, it doesn't.

Can you show your executios of commands executions which I've provided?
Did you try to reproduce what I've done? 🤔

**Comment by mlschroe** [comment]
Your "test-define" example is also not working like you expect. If you add
```
%define test 2
```
at the top of the specfile and add "%test" to your echo statement, you'll see that it echos "2" even if you call rpmbuild with `--define "test 1"`.

Basically --define and --undefine work similar to adding define/undefine to the top of the spec file. So it is not possible to change a %define that comes later on.
